ipwb/replay.py

```diff
@@ -179,6 +179,126 @@ def showMementosForURIRs_sansJS():
     return redirect('/memento/*/' + urir, code=301)
 
 
+def lookup_keys(surt):
+    keyre = re.compile(r"(.+)([,/]).+")
+    key = surt.split("?")[0].strip("/")
+    keys = [key, f"{key}/*"]
+    while "," in key:
+        try:
+            m = keyre.match(key)
+            keys.append(f"{m[1]}{m[2]}*")
+            key = m[1]
+        except Exception as e:
+            key = key.strip("/,")
+
+    return keys
+
+
+def bin_search(iter, key, datetime=None):
+    # Read line encompassing current position
+    ln = iter.readline()
+
+    while ipwbUtils.isCDXJMetadataRecord(ln):
+        ln = iter.readline()
+
+    surtk, rest = ln.split(maxsplit=1)
+    if key == surtk:
+        return [ln]
+
+    # If further searching required...
+    left = 0
+    # Go to end of seek stream
+    iter.seek(0, 2)
+    right = iter.tell()  # Current position
+
+    lines = set()  # Prevents dupes
+    while (right - left > 1):
+        mid = (right + left) // 2
+        iter.seek(mid)
+        iter.readline()  # Purge rest of current line
+        line = iter.readline()  # Read the next full line
+
+        if len(line) == 0:
+            right = mid
+            continue
+
+        surtk, rest = line.split(maxsplit=1)
+        datetimeK = rest.split()[0].decode()
+        if surtk[-1:] == b'/':
+            surtk = surtk[0:-1]
+
+        # TODO: find a more elegant way for comparison than manually
+        # trimming the last two chars off of the surtk
+
+        if key == surtk:
+            if datetime and datetime != datetimeK:
```

Comment:
I think you wanted to check for equality, not otherwise?

```suggestion
            if datetime and datetime == datetimeK:
```

Reply:
Yes, you are correct. Odd that this is the inverse logic but still worked. I will look into it further.

Reply:
Turns out it was not working and giving us false positives of URI-Ms resulting in the wrong memento being returned. Fixed in a603cf5.